### DISCUSSION TITLE
Support static page generation on demand at runtime (for SSG/ISR)

### DISCUSSION THREAD

**Discussion by suchardao** [vercel/next.js#33531] [Ideas]
### Describe the feature you'd like to request

Add support for generating static pages (without slug) on demand = at runtime (not only at build time).

Currently NextJS support generation on demand only for pages which implements `getStaticPaths` (page with slug). But static page without slug (like "pages/index.js" for homepage, "404.js", etc.) cannot implement `getStaticPaths`, so they need to be generated at build time. But in many cases this is not enough. For example when you need to support multiple instances of NexJS application and each instance has its own endpoints for data, cms etc. In this case you cannot fetch data in `getStaticProps` at build time, it's necessary to pre-generate pages at runtime in each instance.

**Comment by ochodek** [comment]
I agree. It would be great! F.e. pages are pre-generated for DEV environment. Why it is "forced" to rebuild the app for STAGE and PROD separately (also multiple languages - per domain)? Or is it possible somehow at this time?

**Reply by justmejulian** [reply]
We currently have the same problem.
Our solution is to remove the generated html pages after the build. This forces them to be rebuilt on first request to said page.
After that they remain static and are served as such.

```bash
❯ npm run build
...

❯ rm .next/server/app/*.html
...

❯ npm run start
```

This is and feels very hacky. But it gives us pre-generate pages at runtime.

**Comment by lihan** [comment]
My use case is that there are thousands of pages stored in the database and would like to generate static version when they are requested. Because during build time, the compile time would be very long to compile thousands of pages.

Is this the same feature requested?

**Reply by suchardao** [reply]
Yes, it will solve your problem. The feature is about the possibility to generate static pages at runtime instead of build time.

**Reply by ollieayre-oc** [reply]
I've been looking into this for a similar reason (to many pages to render at build). It seems that you can achieve this behaviour with this setup:

```
getStaticPaths() {
    return {
      fallback: 'blocking', //<<< forces a static render if one hasn't already been performed.
      paths: [] //<<< Will result in no pages being rendered during a build
    };
}
```

If you need continuous revalidation after the initial render you can set the `revalidate` property in `getStaticProps`.

**Comment by rhyshalsey** [comment]
I'd also like this feature. In our deployment system, our environment variables are injected at runtime, so any page generated with getStaticProps at build time will not have the right environment variables. Ideally, in my case, there would be a way to request a build on the first request during runtime and then all subsequent requests could be cached. 

At the moment, putting 1 second under the "revalidate" option is a kind of solution for us and a bad one. It defeats the purpose of static generation.

**Comment by Ryuurock** [comment]
+1，my css file is generated  at runtime

**Comment by Mouaz** [comment]
+1 in case of websites with thousands rarely accessed pages

**Comment by tomihq** [comment]
+1
